src/pages/ProfilesExplorerView/domain/variables/FiltersVariable/filters-ops.tsx

```diff
@@ -20,16 +20,152 @@ export const convertPyroscopeToVariableFilter = (filter: CompleteFilter): AdHocV
   };
 };
 
-export const addFilter = (model: FiltersVariable, filter: AdHocVariableFilter) => {
+const includeExcludeValue = (
+  operator: string,
+  model: FiltersVariable,
+  filter: AdHocVariableFilter
+): { done: boolean; found?: AdHocVariableFilter } => {
   const found = model.state.filters.find((f) => f.key === filter.key);
 
-  if (found) {
-    found.value = filter.value;
-    model.setState({ filters: [...model.state.filters] });
+  if (!found) {
+    // create a new filter
+    const newFilter = { ...filter, operator: operator }; // will be translated to an "in"/"not in" operator in the UI
+    model.setState({
+      filters: [...model.state.filters, newFilter],
+    });
+    return { done: true };
+  }
+
+  // is there already a filter with either a regex or an "in"/"not in" operator?
+  if (found.operator !== operator) {
+    return { done: false, found };
+  }
+
+  // is it a filter with a regex?
+  if (REGEX_CHARS_REGEX.test(found.value)) {
+    // yes, we replace it by a new filter with an "in"/"not in" operator and the single filter value
+    const newFilter = { ...found, value: filter.value };
+    model.setState({
+      filters: [...model.state.filters.filter((f) => f.key !== filter.key), newFilter],
+    });
+    return { done: true };
+  }
+
+  // we're handling a filter with an "in"/"not in" operator
+  const values = found.value.split('|');
+
+  // does it already include the new value?
+  if (!values.includes(filter.value)) {
+    // no, we include it
+    const newFilter = { ...found, value: [...values, filter.value].join('|') };
+    model.setState({
+      filters: [...model.state.filters.filter((f) => f.key !== filter.key), newFilter],
+    });
+    return { done: true };
+  }
+
+  return { done: true };
+};
+
+export const includeInFilter = (model: FiltersVariable, filter: AdHocVariableFilter) => {
```

Comment:
If this draft PR makes sense, this code will need unit tests. Same below for `excludeFromFilter `.